### PR TITLE
Fix #2498: IR-check the scope of local identifiers.

### LINKED ISSUE #2498: IR-check that identifiers for labels and local variables are unique per method

### PR DESCRIPTION
**Pull request by sjrd**
* Identifiers for labels must be unique per method.
* Identifiers for local variables must have non-overlapping scopes.

### CODE REVIEW

**Comment by sjrd** [comment]
Review by @gzm0

**Comment by sjrd** [comment]
Updated with disabling the check in exported methods.

**Comment by gzm0** [comment]
So I'm assuming you will change this again. Or do you want to get this in and change it again?

**Comment by sjrd** [comment]
I'll change this again, yes. So on standby for now.

**Comment by sjrd** [comment]
Updated.

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/checker/IRChecker.scala [hunk 1]
```diff
@@ -828,33 +850,41 @@ private final class IRChecker(unit: LinkingUnit, logger: Logger) {
...
+          }
+
+          for (ParamDef(name, ptpe, mutable, rest) <- params) {
+            if (ptpe == NoType)
+              reportError(s"Parameter $name has type NoType")
+            else if (ptpe != AnyType)
+              reportError(s"Closure parameter $name has type $ptpe instead of any")
+            if (rest)
+              reportError(s"Closure parameter $name cannot be a rest parameter")
+          }
```
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/checker/IRChecker.scala]: Might make sense to revert this. It's not required anymore, is it?

**Comment by gzm0** [comment]
That's all.

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/checker/IRChecker.scala (hunk 1 above)
Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/checker/IRChecker.scala]: It's still required for label identifiers. Besides, it just makes sense.

Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/checker/IRChecker.scala]: Ok, fair enough. But ParamDef can't really have a label in it, can it? :)

**Comment by gzm0** [comment]
LGTM

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/checker/IRChecker.scala (hunk 1 above)
Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/checker/IRChecker.scala]: `ParamDef`, no. But `captureValues`, definitely.

**Pull request merged**